Add basic support for group replication

ejortegau opened the pull request:
Related issue: https://github.com/openark/orchestrator/issues/1179
### Description

This PR adds some initial support for group replication in orchestrator.

ToDo:
- [X] contributed code is using same conventions as original code
- [X] code is formatted via gofmt (please avoid goimports)
- [X] code is built via ./build.sh
- [x] code is tested via `go test ./go/...`
- [x] relevant documentation has been updated.
- [x] upgrade from previous version (`master` branch) is successful

Assume a 3 member group, plus an async slave replicating from one of them. Orchestrator shows all three members as separate clusters, as it does not understand group replication and there is no traditional replication configured in any group member. The async slave is shown as a slave in the cluster whose master is the group member the slave is set up to replicate from.

Orchestrator also shows the async slave with a problem of errant GTIDs.

With this  PR, some basic support for single primary replication groups is added. So, a replication group shown like this in mysqlshell:
```json
{
    "clusterName": "test_cluster", 
    "defaultReplicaSet": {
        "name": "default", 
        "primary": "127.0.0.1:5306", 
        "ssl": "REQUIRED", 
        "status": "OK", 
        "statusText": "Cluster is ONLINE and can tolerate up to ONE failure.", 
        "topology": {
            "127.0.0.1:4306": {
                "address": "127.0.0.1:4306", 
                "mode": "R/O", 
                "readReplicas": {}, 
                "replicationLag": null, 
                "role": "HA", 
                "status": "ONLINE", 
                "version": "8.0.20"
            }, 
            "127.0.0.1:5306": {
                "address": "127.0.0.1:5306", 
                "mode": "R/W", 
                "readReplicas": {}, 
                "replicationLag": null, 
                "role": "HA", 
                "status": "ONLINE", 
                "version": "8.0.20"
            }, 
            "127.0.0.1:6306": {
                "address": "127.0.0.1:6306", 
                "mode": "R/O", 
                "readReplicas": {}, 
                "replicationLag": null, 
                "role": "HA", 
                "status": "ONLINE", 
                "version": "8.0.20"
            }
        }, 
        "topologyMode": "Single-Primary"
    }, 
    "groupInformationSourceMember": "127.0.0.1:5306"
}
```
Is now understood by orchestrator, as two group secondary group members that replicate from the primary. In addition, the async/replica continues to be shown normally:

![image](https://user-images.githubusercontent.com/5791035/83362216-c80a7200-a38f-11ea-96ee-078d1a7d27f2.png)

Notice also that the errant GTID problem is not shown.

The UI also shows group membership information, including:
* An icon showing whether the instance is a group member or not. A different text style is used for
  the primary and secondary instances so that they can be easily  identified.
* Hovering over the icon shows the instance state and role in the replication group.
* Group members that are not online are shown as with problems, as shown below:

![image](https://user-images.githubusercontent.com/5791035/83426417-60a3ff00-a42f-11ea-9e5b-3efd10848211.png)

On top of this, certain replication operations are prevented from taking place. Namely, attempting to relocate group secondaries to replicate from outside the group, since they actually replicate from the primary, as well as attempting to set-up a group primary to replicate from a secondary.

ejortegau committed:
Initial support for group replication.

* Add group replication fields to database_instance table.
* Add group replication fields to Instance, and a function to evaluate whether
  a host is member of a replication group or not.
* Modifued `Instance.IsMaster()` to not consider a host as a Master if it is a
  secondary member of a replication group.
* Defined constants for group replication member roles  and states.
* `ReadTopologyInstanceBufferable()` now attempts to populate group replication
  fields in Oracle MySQL >= 5.7
* `DiscoverInstance()` not attempts to discover not only an instance's slave
  hosts but also the member of its replication group

ejortegau committed:
Initial support for group replication.

* Adde methods to evaluate whether an instance is either the primary or
  secondary of a replication group.
* Replication group members with secondary role now get their cluster
  attributes populated from the group's primary. The group's primary gets them,
  in turn, from its async/semi-sync master (if any).
* Group replication attributes from instances are now read from the backend DB,
  and therefore, are now correctly returned by the API.

ejortegau committed:
Initial support for group replication.

* Replication group secondary members now are shown in the UI as replicating
  from the group primary.

ejortegau committed:
Initial support for group replication.

* Stop considering GTIDs coming from the group as errant GTIDs. While at it,
  fix tiny typo.

ejortegau committed:
Initial support for group replication.

* Disallow setting a group primary to replicate from a secondary.

ejortegau committed:
Initial support for group replication.

* Split DB migration adding multiple columns in a single ALTER statement to
  make CI happy.

ejortegau commented:
cc @dveeden , @sjmudd , @luisyonaldo

ejortegau committed:
Initial support for group replication.

* Add instance icon showing whether the host is a replication group member as
  well as its group role and state.

ejortegau committed:
Initial support for group replication.

* Replication group members that have been expelled from the group are still
  identified as part of the same cluster and shown in the topology, instead of
  appearing as separate chains.

ejortegau committed:
Initial support for group replication.

* Group members that are not ONLINE are now exposed as having problems through
  the API.
* The web UI now shows replication group members that have been expelled by the
  group majority.
* Group members that are not ONLINE are now shown in the web UI in either red
  or orange depending on their state in the group (orange when RECOVERING, red
  when OFFLINE or in ERROR).

ejortegau committed:
Initial support for group replication.

* Update docs to include information on partial support for GR.

dveeden commented:
I think group replication is an important feature for Orchestrator as it looks like this is at the core of the long term strategy of Oracle MySQL. Group Replication is a core component of InnoDB Cluster.

Note that *some* info about group replication is available in the same way as normal replication. There is nothing in `SHOW SLAVE STATUS`, but `performance_schema.replication_connection_status` etc show a `group_replication_applier` and a `group_replication_recovery` channel. Other newer / advanced replication features like multi source replication and parallel replication use the same performance_schema tables. So I think that where possible these should be used. For GR there are additional tables like `replication_group_members`.

@ejortegau, Could you squash your commits? Or are there multiple on purpose?

ejortegau commented:
> @ejortegau, Could you squash your commits? Or are there multiple on purpose?

There are multiple on purpose, as I have worked on it and added more and more support so far. I figure they would/could be squashed once the merge is accepted.

dveeden commented on go/db/generate_patches.go:
```diff
@@ -576,4 +576,40 @@ var generateSQLPatches = []string{
+	`
+		ALTER TABLE
+			database_instance
+			ADD COLUMN replication_group_name VARCHAR(64) CHARACTER SET ascii NOT NULL DEFAULT '' AFTER gtid_mode
```
So this is a UUID?
https://dev.mysql.com/doc/refman/5.7/en/group-replication-options.html#sysvar_group_replication_group_name

ejortegau replied:
Yes, it is a UUID. I considered storing it with a binary, but chose not to in order to be consistent with how other UUIDs (e.g. `server_uuid`) are being stored in the orchestrator database.

dveeden commented on go/inst/instance_dao.go:
```diff
@@ -478,6 +491,10 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool,
 					_ = db.QueryRow("select count(*) > 0 and MAX(User_name) != '' from mysql.slave_master_info").Scan(&instance.ReplicationCredentialsAvailable)
 				}
+				// Populate GR information for the instance in Oracle MySQL 5.7+
+				if !instance.IsSmallerMajorVersionByString("5.7") {
```
What about MariaDB? Should we exclude that?

ejortegau replied:
This is already in an if block with this:
```go
if (instance.IsOracleMySQL() || instance.IsPercona()) && !instance.IsSmallerMajorVersionByString("5.6") {
```
So MariaDB is  already excluded.

dveeden commented:
Would this work ok with a group in multi primary mode?

https://dev.mysql.com/doc/refman/8.0/en/group-replication-multi-primary-mode.html

ejortegau committed:
Initial support for group replication.

* Update docs to include information on partial support for GR.

luisyonaldo commented on go/inst/instance_dao.go:
```diff
@@ -478,6 +491,10 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool,
+				// Populate GR information for the instance in Oracle MySQL 5.7+
+				if !instance.IsSmallerMajorVersionByString("5.7") {
+					PopulateGroupReplicationInformation(instance, db)
+				}
```
add a new `waitGroup` and do this inside a new goroutine

ejortegau reviewed (commented):
> Would this work ok with a group in multi primary mode?
> 
> https://dev.mysql.com/doc/refman/8.0/en/group-replication-multi-primary-mode.html

It will not. I have missed pushing an update to the FAQ, indicating this. For now, this only works with single primary mode.

luisyonaldo commented on go/inst/instance_dao.go:
```diff
@@ -3068,6 +3155,79 @@ func FigureInstanceKey(instanceKey *InstanceKey, thisInstanceKey *InstanceKey) (
+// (if any). This method swallows up any errors, interpreting them as if the host does not support group replication
+// (e.g. because the group replication plugin is not loaded, therefore the corresponding performance schema tables and
+// global variables are not known to the server.
+func PopulateGroupReplicationInformation(instance *Instance, db *sql.DB) {
```
I'm not sure ignoring the error is the best way here, I would rather catch the error and ignore only the expected errors (no group replication plugin, etc), instead of ignoring all errors (which might bypass some error that we would like to see)

shlomi-noach replied:
Agree

shlomi-noach replied:
Is it not possible to validate whether this host at all supports group replication, and based on that finding to invoke/skip this function?

ejortegau replied:
Yes, we can check whether the plugin is installed and active, but this requires an extra query. I will modify this to check the error codes from the query and only ignore them if they are either because the group replication global variables or performance schema tables are not known, which would mean that the host does not support group replication.

ejortegau replied:
Done.

luisyonaldo commented on go/inst/instance_dao.go:
```diff
@@ -744,6 +761,10 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool,
+		// wait for routines to finish before reading cluster attributes. This seems to be needed, as,
+		// otherwise, detecting whether a host is a replication group secondary to set up cluster attributes from its primary
+		// does not work. ToDo: Is there a better way to sync this?
+		waitGroup.Wait()
```
this seems wrong, why do you need to wait here?

ejortegau replied:
Without this, it would sometimes think that a group member was not actually a group member, because the `ReplicationGroupName` attribute was not populated yet. Hence I added the wait.

shlomi-noach replied:
This will hurt parallelism. I see the dependency problem. Let me see if moving `ReadInstanceClusterAttributes` itself can  break dependencies.

luisyonaldo replied:
this shouldn't be in a wait group since it has to be applied in sequential order, same as how `show slave status` query is read.

I would put this after `show slave status` code:
```go
instance.ReplicationIOThreadState = ReplicationThreadStateNoThread
instance.ReplicationSQLThreadState = ReplicationThreadStateNoThread
err = sqlutils.QueryRowsMap(db, "show slave status", func(m sqlutils.RowMap) error {
...
if err != nil {
	goto Cleanup
}
if instance.IsOracleMySQL() && !instance.IsSmallerMajorVersionByString("5.7") {
	err = PopulateGroupReplicationInformation(instance, db)
        if err != nil {
	        goto Cleanup
        }
}
```

ejortegau replied:
So this took me a while to fully understand, but the problem comes from the fact that in order to populate  GR attributes of the instance, the Server UUID needs to be known. This UUID is obtained asynchronously, so I added a wait group for it and defer execution of the `PopulateGroupReplicationInformation` until after the UUID is known. `PopulateGroupReplicationInformation` itself is run sequentially, as per @luisyonaldo 's suggestion.

luisyonaldo commented on go/inst/instance_dao.go:
```diff
@@ -3068,6 +3155,79 @@ func FigureInstanceKey(instanceKey *InstanceKey, thisInstanceKey *InstanceKey) (
 	return figuredKey, nil
 }
 
+// Obtain information about Group Replication  for this host as well as other hosts who are members of the same group
```
`// PopulateGroupReplicationInformation  obtain information about ... `

luisyonaldo commented on go/inst/instance_dao.go:
```diff
@@ -3068,6 +3155,79 @@ func FigureInstanceKey(instanceKey *InstanceKey, thisInstanceKey *InstanceKey) (
+  	FROM
+		performance_schema.replication_group_members
+  	 `
+	rows, err := db.Query(q)
```
why not `db.QueryRow`? https://golang.org/pkg/database/sql/#DB.QueryRow 
 I think is being used everywhere already, otherwise please add `defer rows.Close()`

ejortegau replied:
Added the defer `rows.Close()`

ejortegau committed:
Initial support for Group Replication

* Fix unit tests that were broken due to addition of new columns to
  `database_instance` table.

ejortegau committed:
Initial support for Group Replication

* Address some MR comments and failing test.

ejortegau committed:
Initial support for Group Replication

* Revert bad change to DB migration.
* Fix SIGSEV introduced by bad placement of `defer rows.Close()`.

ejortegau committed:
Initial support for Group Replication

* Fix integration testing that was failing due to missing value on `INSERT`
  statement for non nullable column `replication_group_members` of table
  `database_instance` which has no default.

dveeden commented:
For groups we should draw a line around them and maybe show the name of the group.
This is true for Group Replication, NDB and maybe PXC.

With D3.js this can be done with `d3.geom.hull`. Example in g1 in the image.

The `d3.layout.tree()` that is used doesn't allow more complex topologies like groups, multi source etc. This *could* be fixed by moving to `d3.layout.force()`. Example in g2 in the image.

I also notice that D3.js v3 is used, which is two versions behind the latest major release (D3.js v5).

![Screenshot from 2020-06-05 15-21-07](https://user-images.githubusercontent.com/1272980/83881217-994e1c00-a740-11ea-81c8-e9b23aba5816.png)

The code for the example is in:
https://gist.github.com/dveeden/8a5cc4c1230f0442bb8b74c8e0eecad7

Some of this could be done outside of this PR, but we should make sure that the nodelist is populated with group info to allow the frontend part to be done later on.

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -204,6 +222,21 @@ func (this *Instance) IsNDB() bool {
+}
+
+func (this *Instance) IsReplicationGroupPrimary() bool {
+	return this.IsReplicationGroupMember() && this.ReplicationGroupPrimaryKey == this.Key
```
Please use `Equals()`
```suggestion
	return this.IsReplicationGroupMember() && this.ReplicationGroupPrimaryKey.Equals(&this.Key)
```

ejortegau replied:
Fixed

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -204,6 +222,21 @@ func (this *Instance) IsNDB() bool {
+}
+
+func (this *Instance) IsReplicationGroupSecondary() bool {
+	return this.IsReplicationGroupMember() && this.ReplicationGroupPrimaryKey != this.Key
```
```suggestion
	return this.IsReplicationGroupMember() && !this.ReplicationGroupPrimaryKey.Equals(&this.Key)
```

ejortegau replied:
Fixed.

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -271,7 +304,10 @@ func (this *Instance) IsReplica() bool {
+	return !this.IsReplica() &&
+		((this.IsReplicationGroupMember() &&
+			this.ReplicationGroupMemberRole != GroupReplicationMemberRoleSecondary) ||
+			!this.IsReplicationGroupMember())
```
too much `&& || (( ))` complexity. I can hardly parse.
```suggestion
        if this.IsReplica() {
          return false
        }
	if !this.IsReplicationGroupMember() {
	  return false
	}
	if this.IsReplicationGroupMember() && this.ReplicationGroupMemberRole == GroupReplicationMemberRoleSecondary {
	  return false
	}
	return true
```

ejortegau replied:
Fixed

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -121,13 +121,31 @@ type Instance struct {
+	ReplicationGroupMembers InstanceKeyMap
+
+	// Primary of the replication group
+	ReplicationGroupPrimaryKey InstanceKey
```
Because `PrimaryKey` is a term in databases, let's use `PrimaryInstanceKey`
```suggestion
	ReplicationGroupPrimaryInstanceKey InstanceKey
```

ejortegau replied:
Fixed

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -866,12 +887,15 @@ Cleanup:
 		}
+		// Add replication group ancestry UUID as well. Otherwise, Orchestrator thinks there are errant GTIDs in group
+		// members and its slaves, even though they are not.
+		instance.AncestryUUID = fmt.Sprintf("%s,%s", instance.AncestryUUID, instance.ReplicationGroupName)
```
is `ReplicationGroupName` a `UUID`?

ejortegau replied:
Indeed, it is, see https://dev.mysql.com/doc/refman/5.7/en/group-replication-options.html#sysvar_group_replication_group_name

shlomi-noach reviewed (changes requested):
Only an initial, partial review. I'll need to revisit.

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -3068,6 +3155,79 @@ func FigureInstanceKey(instanceKey *InstanceKey, thisInstanceKey *InstanceKey) (
+		@@global.group_replication_group_name,
+		@@global.group_replication_single_primary_mode
+  	FROM
+		performance_schema.replication_group_members
```
Does this require any special privileges? (I don't know)

ejortegau replied:
AFAIK just regular `SELECT`  on the relevant `performance_schema` tables. I will check to validate this assumption.

ejortegau replied:
Confirmed, a user with only `SELECT` privilege in `performance_schema.replication_group_members` can run the query.

shlomi-noach replied:
So, it _does_ require special privileges. Can you please document this in https://github.com/openark/orchestrator/blob/master/docs/configuration-discovery-basic.md ?

ejortegau replied:
Done

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -2695,10 +2695,19 @@ func RelocateBelow(instanceKey, otherKey *InstanceKey) (*Instance, error) {
 	}
+	// Relocation of group secondaries makes no sense, group secondaries, by definition, always replicate from the group
+	// primary
+	if instance.IsReplicationGroupSecondary() {
```
I'm thinking the same logic needs to be applied in `MoveBelow`, `MoveBelowGTID`, `MakeCoMaster`, `Repoint`, `TakeMaster`, `MatchBelow`, `MatchUp` functions.

ejortegau replied:
Will apply there.

ejortegau replied:
Fixed

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -600,8 +600,17 @@ function normalizeInstance(instance) {
+  // Setting the masterId to the group primary is what allows us to visualize group secondary members as replicating
+  // from the group primary.
+  if (instance.ReplicationGroupName != "" && (instance.ReplicationGroupMemberRole == "SECONDARY" || instance.ReplicationGroupMemberRole == ""))
+    masterKey = instance.ReplicationGroupPrimaryKey
```
```suggestion
    masterKey = instance.ReplicationGroupPrimaryKey;
```

ejortegau replied:
Fixed

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -600,8 +600,17 @@ function normalizeInstance(instance) {
+  if (instance.ReplicationGroupName != "" && (instance.ReplicationGroupMemberRole == "SECONDARY" || instance.ReplicationGroupMemberRole == ""))
+    masterKey = instance.ReplicationGroupPrimaryKey
+  else
+    masterKey = instance.MasterKey
```
```suggestion
    masterKey = instance.MasterKey;
```

ejortegau replied:
Fixed

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -600,8 +600,17 @@ function normalizeInstance(instance) {
+  else
+    masterKey = instance.MasterKey
+  instance.masterId = getInstanceId(masterKey.Hostname,
+    masterKey.Port);
```
```suggestion
  instance.masterId = getInstanceId(masterKey.Hostname, masterKey.Port);
```

ejortegau replied:
Moved into a single line instead

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -1182,6 +1250,10 @@ func readInstanceRow(m sqlutils.RowMap) *Instance {
+	// Group replication problems
+	if instance.ReplicationGroupName != "" && instance.ReplicationGroupMemberState != "ONLINE" {
+		instance.Problems = append(instance.Problems, "group_replication_member_not_online")
+	}
```
Can we "upgrade" this to be analyzed in `instance.go`?

ejortegau replied:
I can move it there, though I thought this was a better place for it. From what I see, `instance.go` does not populate problems anywhere; instead, they seem to be populated here.

shlomi-noach reviewed (changes requested):
Some further comments. Review is not yet complete.

Thank you!

ejortegau committed:
Initial support for Group Replication - Addressing PR comments

* Instance attribute renamed from `ReplicationGroupPrimaryKey` to
  `ReplicationGroupPrimaryInstanceKey`.
* Use `Instance.Equals()`.
* Easier to read `Instance.IsMaster()`.
* Don't blindly assume that any error coming from executing a query to find out
  group replication attributes for an instance means that the instance does not
  support group replication. Instead, check error codes to determine whether
  the error comes from that or from something else, and have different behavior
  depending on the answer.
* Fix issue leading to random detection of group secondaries as not group
  members, coming from the fact that, while attempting to find the member in
  the records of `performance_schema.replication_group_members`, the required
  attribute for server UUID had not yet been populated. This should also reduce
  the loss of parallelism previously present, as new WaitGroup only waits for
  determining the server UUID instead of for all standing routines to finish.
* Prevent replication refactoring operations from taking place for group
  secondaries since they make no sense for them.
* Group secondaries with replicas under them no longer are shown as with a
  `NoWriteableMasterStructureWarning` problem.

ejortegau committed:
Initial support for Group Replication - Addressing PR comments

* Remove no longer needed `WaitGroup.wait()`.

ejortegau committed:
Initial support for Group Replication - Addressing PR comments

* Fix formatting issues.

ejortegau committed:
Initial support for Group Replication

* Re-order of GR constants to have most frequently used ones above.

shlomi-noach committed:
Merge branch 'master' into ejortegau/add_basic_support_for_group_replication

shlomi-noach committed:
Merge branch 'master' into ejortegau/add_basic_support_for_group_replication

eurrego-at-booking committed:
Merge branch 'master' into ejortegau/add_basic_support_for_group_replication

eurrego-at-booking committed:
Add info on grants needed for GR support

shlomi-noach commented:
Sorry for the delays. I hope to be able to re-review and merge early next week (Sunday/Monday)

shlomi-noach committed:
formatting

shlomi-noach commented on go/inst/analysis_dao.go:
```diff
@@ -160,11 +160,19 @@ func GetReplicationAnalysis(clusterName string, hints *ReplicationAnalysisHints)
+			AND (
+				master_instance.replication_group_name = ''
+				OR master_instance.replication_group_member_role = 'PRIMARY'
+			)
```
:+1:

shlomi-noach committed:
formatting

shlomi-noach commented on docs/faq.md:
```diff
@@ -62,7 +62,21 @@ No.
+
+No support has been added (yet) to handling group member failure. If all you have is a single replication group, this is fine, because you don't need it; the group will handle all failures as long as it can secure a majority.
+
+If, however, you have the primary of a group as a replica to another instance; or you have replicas under your group
```
:+1:

shlomi-noach commented on go/db/generate_patches.go:
```diff
@@ -580,5 +580,41 @@ var generateSQLPatches = []string{
+	`
+		ALTER TABLE
+		database_instance
+			ADD COLUMN replication_group_primary_port smallint(5) unsigned NOT NULL DEFAULT 0 AFTER replication_group_primary_host
```
:+1:

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -129,13 +129,32 @@ type Instance struct {
+	ReplicationGroupMembers InstanceKeyMap
+
+	// Primary of the replication group
+	ReplicationGroupPrimaryInstanceKey InstanceKey
```
:+1:

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -129,13 +129,32 @@ type Instance struct {
-		Problems: []string{},
+		Replicas:                make(map[InstanceKey]bool),
+		ReplicationGroupMembers: make(map[InstanceKey]bool),
+		Problems:                []string{},
```
:+1:

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -228,6 +247,21 @@ func (this *Instance) IsNDB() bool {
+
+func (this *Instance) IsReplicationGroupSecondary() bool {
+	return this.IsReplicationGroupMember() && !this.ReplicationGroupPrimaryInstanceKey.Equals(&this.Key)
+}
```
:+1:

shlomi-noach committed:
simplify nesting

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -293,9 +327,27 @@ func (this *Instance) IsReplica() bool {
+				return false
+			}
+		}
+	}
```
I'm replacing this code:
```golang
func (this *Instance) IsMaster() bool {
	// If traditional replication is configured, it is for sure not a master
	if this.IsReplica() {
		return false
	} else {
		// If traditional replication is not configured, and it is also not part of a replication group, this host is
		// a master
		if !this.IsReplicationGroupMember() {
			return true
		} else {
			// If traditional replication is not configured, and this host is part of a group, it is only considered a
			// master if it has the role of group Primary. Otherwise it is not a master.
			if this.ReplicationGroupMemberRole == GroupReplicationMemberRolePrimary {
				return true
			} else {
				return false
			}
		}
	}
}
```
with simplified nesting:
```golang
func (this *Instance) IsMaster() bool {
	// If traditional replication is configured, it is for sure not a master
	if this.IsReplica() {
		return false
	}
	// If traditional replication is not configured, and it is also not part of a replication group, this host is
	// a master
	if !this.IsReplicationGroupMember() {
		return true
	}
	// If traditional replication is not configured, and this host is part of a group, it is only considered a
	// master if it has the role of group Primary. Otherwise it is not a master.
	if this.ReplicationGroupMemberRole == GroupReplicationMemberRolePrimary {
		return true
	}
	return false
}
```

shlomi-noach commented on go/inst/instance.go:
```diff
@@ -358,6 +407,11 @@ func (this *Instance) AddReplicaKey(replicaKey *InstanceKey) {
+// AddGroupMemberKey adds a group member to the list of this instance's group members.
+func (this *Instance) AddGroupMemberKey(groupMemberKey *InstanceKey) {
+	this.ReplicationGroupMembers.AddKey(*groupMemberKey)
+}
```
:+1:

shlomi-noach committed:
formatting

shlomi-noach committed:
formatting

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -485,6 +485,11 @@ func MoveBelow(instanceKey, siblingKey *InstanceKey) (*Instance, error) {
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, log.Errorf("MoveBelow: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -656,6 +661,11 @@ func MoveBelowGTID(instanceKey, otherKey *InstanceKey) (*Instance, error) {
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, log.Errorf("MoveBelowGTID: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -766,7 +776,11 @@ func Repoint(instanceKey *InstanceKey, masterKey *InstanceKey, gtidHint Operatio
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, fmt.Errorf("repoint: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -926,6 +940,11 @@ func MakeCoMaster(instanceKey *InstanceKey) (*Instance, error) {
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, fmt.Errorf("MakeCoMaster: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -1522,6 +1541,11 @@ func MatchBelow(instanceKey, otherKey *InstanceKey, requireInstanceMaintenance b
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, nil, fmt.Errorf("MatchBelow: %+v is a secondary replication group member, hence, it cannot be relocated", *instanceKey)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -1731,6 +1755,11 @@ func TakeMaster(instanceKey *InstanceKey, allowTakingCoMaster bool) (*Instance,
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, fmt.Errorf("takeMaster: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -2026,6 +2055,11 @@ func MatchUp(instanceKey *InstanceKey, requireInstanceMaintenance bool) (*Instan
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, nil, fmt.Errorf("MatchUp: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -2695,10 +2729,19 @@ func RelocateBelow(instanceKey, otherKey *InstanceKey) (*Instance, error) {
+	// primary
+	if instance.IsReplicationGroupSecondary() {
+		return instance, log.Errorf("relocate: %+v is a secondary replication group member, hence, it cannot be relocated", instance.Key)
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_topology.go:
```diff
@@ -2695,10 +2729,19 @@ func RelocateBelow(instanceKey, otherKey *InstanceKey) (*Instance, error) {
+	// Disallow setting up a group primary to replicate from a group secondary
+	if instance.IsReplicationGroupPrimary() && other.ReplicationGroupName == instance.ReplicationGroupName {
+		return instance, log.Errorf("relocate: Setting a group primary to replicate from another member of its group is disallowed")
+	}
```
:+1:

shlomi-noach commented on go/logic/orchestrator.go:
```diff
@@ -279,8 +279,8 @@ func DiscoverInstance(instanceKey inst.InstanceKey) {
-	// Investigate replicas:
-	for _, replicaKey := range instance.Replicas.GetInstanceKeys() {
+	// Investigate replicas and members of the same replication group:
+	for _, replicaKey := range append(instance.ReplicationGroupMembers.GetInstanceKeys(), instance.Replicas.GetInstanceKeys()...) {
```
:+1:

shlomi-noach commented on resources/public/js/cluster-analysis-shared.js:
```diff
@@ -46,5 +46,9 @@ var errorMapping = {
+	},
+	"group_replication_member_not_online": {
+		"badge": "label-danger",
+		"description": "Replication group member is not ONLINE"
```
:+1:

shlomi-noach replied:
:+1:

shlomi-noach committed:
formatting

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -600,8 +600,16 @@ function normalizeInstance(instance) {
+  if (instance.ReplicationGroupName != "" && (instance.ReplicationGroupMemberRole == "SECONDARY" || instance.ReplicationGroupMemberRole == ""))
+    masterKey = instance.ReplicationGroupPrimaryInstanceKey;
+  else
+    masterKey = instance.MasterKey;
```
:+1:

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -665,6 +673,9 @@ function normalizeInstanceProblem(instance) {
   }
+  instance.replicationGroupMemberStateProblem = function() {
+    return instanceProblemIfExists("group_replication_member_not_online")
+  }
```
:+1:

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -691,6 +702,9 @@ function normalizeInstanceProblem(instance) {
     instance.problemOrder = 6;
+  } else if (instance.replicationGroupMemberStateProblem()) {
+    instance.problemDescription = "Replication group member in state " + instance.ReplicationGroupMemberState;
+    instance.problemOrder = 7;
```
:+1:

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -908,6 +922,16 @@ function renderInstanceElement(popoverElement, instance, renderType) {
+        text_style = "";
+      else
+        text_style = "text-muted";
+      popoverElement.find("h3 div.pull-right").prepend('<span class="glyphicon '+ text_style +' glyphicon-tower" title="Group replication '+ instance.ReplicationGroupMemberRole + ' ' + instance.ReplicationGroupMemberState +'"></span> ');
```
:+1:

shlomi-noach commented on resources/public/js/orchestrator.js:
```diff
@@ -940,6 +964,15 @@ function renderInstanceElement(popoverElement, instance, renderType) {
+        default:
+          instance.renderHint = "danger";
+          break;
+      }
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -67,6 +68,29 @@ func (this InstancesByCountReplicas) Less(i, j int) bool {
+	GroupReplicationMemberStateOnline     = "ONLINE"
+	GroupReplicationMemberStateRecovering = "RECOVERING"
+	GroupReplicationMemberStateOffline    = "OFFLINE"
+	GroupReplicationMemberStateError      = "ERROR"
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -67,6 +68,29 @@ func (this InstancesByCountReplicas) Less(i, j int) bool {
+	// replication, at least in the form supported here.
+	1193: true, // ERROR: 1193 (HY000): Unknown system variable 'group_replication_group_name'
+	1146: true, // ERROR: 1146 (42S02): Table 'performance_schema.replication_group_members' doesn't exist
+}
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -580,6 +607,17 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool,
+	// be populated to be able to find this instance's information in performance_schema.replication_group_members by
+	// comparing UUIDs. We could instead resolve the MEMBER_HOST and MEMBER_PORT columns into an InstanceKey and compare
+	// those instead, but this could require external calls for name resolving, whereas comparing UUIDs does not.
+	serverUuidWaitGroup.Wait()
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -580,6 +607,17 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool,
+		if err != nil {
+			goto Cleanup
+		}
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -866,12 +904,15 @@ Cleanup:
 		}
+		// Add replication group ancestry UUID as well. Otherwise, Orchestrator thinks there are errant GTIDs in group
+		// members and its slaves, even though they are not.
+		instance.AncestryUUID = fmt.Sprintf("%s,%s", instance.AncestryUUID, instance.ReplicationGroupName)
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -943,20 +984,45 @@ func ReadClusterAliasOverride(instance *Instance) (err error) {
+	})
+	return err
+}
+
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -943,20 +984,45 @@ func ReadClusterAliasOverride(instance *Instance) (err error) {
-	var masterExecutedGtidSet string
-	masterDataFound := false
+	var masterOrGroupPrimaryExecutedGtidSet string
+	masterOrGroupPrimaryDataFound := false
```
:+1:  to be later all renamed as: `s/masterOrGroupPrimary/primary`

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -969,17 +1035,24 @@ func ReadInstanceClusterAttributes(instance *Instance) (err error) {
+		masterOrGroupPrimaryInstanceKey = instance.ReplicationGroupPrimaryInstanceKey
+	} else {
+		masterOrGroupPrimaryInstanceKey = instance.MasterKey
+	}
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -988,9 +1061,9 @@ func ReadInstanceClusterAttributes(instance *Instance) (err error) {
-		clusterName = masterClusterName
+	if masterOrGroupPrimaryDataFound {
+		replicationDepth = masterOrGroupPrimaryReplicationDepth + 1
+		clusterName = masterOrGroupPrimaryClusterName
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -1015,11 +1088,11 @@ func ReadInstanceClusterAttributes(instance *Instance) (err error) {
 	instance.AncestryUUID = ancestryUUID
-	instance.masterExecutedGtidSet = masterExecutedGtidSet
+	instance.masterExecutedGtidSet = masterOrGroupPrimaryExecutedGtidSet
 	return nil
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -1182,6 +1265,10 @@ func readInstanceRow(m sqlutils.RowMap) *Instance {
+	// Group replication problems
+	if instance.ReplicationGroupName != "" && instance.ReplicationGroupMemberState != GroupReplicationMemberStateOnline {
+		instance.Problems = append(instance.Problems, "group_replication_member_not_online")
+	}
```
:+1:

shlomi-noach committed:
sql syntax

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -2440,6 +2528,13 @@ func mkInsertOdkuForInstances(instances []*Instance, instanceWasActuallyFound bo
+		"replication_group_member_role",
+		"replication_group_members",
+		"replication_group_primary_host",
+		"replication_group_primary_port",
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -2526,6 +2621,13 @@ func mkInsertOdkuForInstances(instances []*Instance, instanceWasActuallyFound bo
+		args = append(args, instance.ReplicationGroupMemberRole)
+		args = append(args, instance.ReplicationGroupMembers.ToJSONString())
+		args = append(args, instance.ReplicationGroupPrimaryInstanceKey.Hostname)
+		args = append(args, instance.ReplicationGroupPrimaryInstanceKey.Port)
```
:+1:

shlomi-noach commented on go/inst/instance_dao.go:
```diff
@@ -3068,6 +3170,92 @@ func FigureInstanceKey(instanceKey *InstanceKey, thisInstanceKey *InstanceKey) (
+				"part of a replication group", instance.Key)
+		}
+	}
+	return nil
```
:+1:

shlomi-noach reviewed (approved):
Approved. I've made a few formatting changes.

earl86 commented:
I find when MGR changed master, orchestrator has no call PostFailoverProcesses hooks.

earl86 commented:
orchestrator  monitor mysql group replication(MGR),when  MGR PRIMARY changed, can orchestrator   update consul key-value info for this MGR cluster?
